src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -269,6 +277,216 @@ private Optional<Pocket> findPocketForReload(ReloadBalanceRequest reloadRequest,
+    }
+
+    private BalanceChangeEventDto revertChargingOperation(Balance balance,
+                                                          RevertBalanceOperationRequest revertRequest,
+                                                          BalanceChangeEventDto chargingOperationToRevert) {
+        String operationUuid = StringUtils.isNotBlank(revertRequest.getUuid())
+            ? revertRequest.getUuid()
+            : UUID.randomUUID().toString();
+
+        BalanceChangeEvent changeEvent = createBalanceChangeEvent(
+            operationUuid,
+            chargingOperationToRevert.getAmountTotal(),
+            Metadata.of(revertRequest.getMetadata()),
+            now(clock),
+            balance,
+            false,
+            REVERT_CHARGING
+        );
+        changeEvent.setRevertOperationId(revertRequest.getUuidToRevert());
+
+        boolean isNeedToReturnAmountToNewPocket = isNeedToReturnAmountToNewPocket(revertRequest.getStartDateTime(), revertRequest.getEndDateTime());
+        log.info("Is need to return amount to new pocket: {}", isNeedToReturnAmountToNewPocket);
+        if (isNeedToReturnAmountToNewPocket) {
+            return revertToNewPocket(balance, revertRequest, chargingOperationToRevert, changeEvent);
+        } else {
+            return revertToOldPockets(chargingOperationToRevert, changeEvent);
+        }
+    }
+
+    private BalanceChangeEventDto revertToNewPocket(Balance balance,
+                                                    RevertBalanceOperationRequest revertRequest,
+                                                    BalanceChangeEventDto chargingOperationToRevert,
+                                                    BalanceChangeEvent changeEvent) {
+        Pocket pocket = new Pocket()
+            .key(randomUUID().toString())
+            .balance(balance)
+            .amount(chargingOperationToRevert.getAmountDelta())
+            .endDateTime(revertRequest.getEndDateTime())
+            .startDateTime(revertRequest.getStartDateTime())
+            .label(revertRequest.getLabel())
+            .metadata(Metadata.of(revertRequest.getMetadata()));
+
+        Pocket savedPocket = pocketRepository.save(pocket);
+        PocketChangeEvent event = createPocketChangeEvent(savedPocket, chargingOperationToRevert.getAmountDelta(), ZERO);
+
+        changeEvent.addPocketChangeEvent(event);
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private BalanceChangeEventDto revertToOldPockets(BalanceChangeEventDto chargingOperationToRevert,
+                                                     BalanceChangeEvent changeEvent) {
+        Map<Long, Pocket> pocketIdToPocketMap = new HashMap<>();
+
+        for (PocketChangeEventDto pocketChangeEventDto : chargingOperationToRevert.getPocketChangeEvents()) {
+            pocketRepository.findOneByIdForUpdate(pocketChangeEventDto.getPocketId())
```

Comment:
copy/paste code

just create get method getPocketForUpdate that throw exception

Reply:
Fixed